### DISCUSSION TITLE
Does MapBox support navigation (turn-by-turn) overlay?

### DISCUSSION THREAD

**Discussion by xrd** [BimmerGestalt/AAIdrive#767] [Q&A]
I'm really happy with the map integration of AAIdrive. It is so cool to see the map rendered in my HUD. However, I am a little confused whether it is working fully, or whether I need to enable something else to get turn by turn navigation.

From other threads here I see the Google Maps does not allow rendering the turn by turn. But I see that MapBox integration has landed in newer builds and "fixes" this. I'm not sure what the fix is.

Is there a way to add turn-by-turn navigation on the HUD in my dashboard? It is great to see a map, but I keep having to glance down at my phone to see when a turn is coming, and this defeats the purpose of the mapping for me. 

Is this something I can enable using a setting? Or, by building the APK myself with a special set of flags? I was an Android developer in a past life; is this something that can be done technically using MapBox (or another integration) and just has not yet been done? Or, is the current implementation the best it can be?

**Comment by hufman** [comment]
Thank you for your compliments! I personally don't feel very proud about the performance, and have not invested more time in the map feature. AAIdrive has not implemented turn-by-turn navigation, this is correct. Mapbox does not have the same policy restriction as Gmap, it just seemed too hard to build myself. Mapbox does offer a full prebuilt Navigation SDK, except it [costs 75x more](https://www.mapbox.com/pricing#navigation-sdk-metered-trips) than [just showing maps](https://www.mapbox.com/pricing#maps-sdks-for-mobile).

The map integration currently implements POI searching, finding a route to a selected POI (or to a destination sent from the phone UI), and showing the route overlay as a line on the map. I believe Mapbox's route API does return a list of segments and their turns, but I couldn't think of a way to locate the (progress along the) current nav segment based on the current GPS point.
The map that is screenshotted and sent to the car is a regular Android Activity, just rendered in an offscreen window, so the next step would be to add Android Views to represent the current step and turn, tied to some logic about showing "keep going straight" until "turn in X distance".

The good news is that this functionality should be relatively standard logic and Android UI development, not tied to the car, so external contributors would be able and welcome to help!

**Reply by xrd** [reply]
Hi. It's great to hear you aren't proud of the mapping because I need to tell you that I absolutely love it. It's so cool, and especially because I know how challenging it is with not just the technical work but the even more challenging license issues.

This is really fascinating to me. Are you saying that if I implemented a different off screen view that I could then use that as the map overlay?

For example, I'm curious if OpenStreetMap could be used instead? It looks like there are many good routing libraries that work on Android, and even some open source applications that (perhaps) a developer could pull a view and reuse it here.

https://wiki.openstreetmap.org/wiki/Routing#:~:text=OpenStreetMap%20data%20includes%20information%20for,turn%20navigation%20app%20for%20iOS).

At any rate, I think this project is so exciting and I love that I can retrofit my car with software of of my own choosing! Thanks so much.

I'll spend some time investigating this and put more in the thread later. If anything I said is an absolute dead end, please let me know.

**Reply by xrd** [reply]
And one extra question: it sounds like Google prohibits using navigation inside an off screen overlay. But I'm curious if you are building your own app (without publishing in the app store) if it is technically possible to do it. I can definitely fork this application and build my own APK and I enjoy doing things like that.

**Reply by hufman** [reply]
Thank you for that compliment, I'm glad you like it!
The window that is shown in the car is a regular subclass of Activity, and [loads a layout file](https://github.com/BimmerGestalt/AAIdrive/blob/14d65d084c66ed20e2984b575464bbc16e4762b2/app/src/mapbox/java/me/hufman/androidautoidrive/carapp/maps/MapboxProjection.kt#L67) just like any other. It is possible to add extra widgets, or hook up databinding, as desired.

Mapbox already uses OpenStreetMap data and routing, and it has the advantage of cloud-hosted maps and traffic information. The OpenStreetMap libraries may have good inspiration for widgets and so on, but it would probably make sense to continue using Mapbox for the navigation and mapping itself.

As for Google Maps, their [API terms of service](https://developers.google.com/maps/terms-20180207) 10.4.c.i disallow creating a substitute service that's similar to the official Google Maps in general, unrelated to publishing to the store. And so, the Gmaps variant of AAIdrive definitely can't have a similar navigation experience, while Mapbox has no such restriction.

I agree with the excitement of building custom apps for the car! While AAIdrive is focused on Android Auto functionality, the [addons](https://github.com/BimmerGestalt/IDriveConnectAddons) have been fun to build too.
